### PR TITLE
Resolves ad-accounts/#913

### PR DESCRIPTION
**Pull request by iambmelt**
See tracking issue:
- https://github.com/AzureAD/ad-accounts-for-android/issues/913

Summary of changes:
- `ApplicationInfo.uid` is now used as a criterion for cache selection
- When performing ['blind lookups'](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/iambmelt/913/common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java#L377-L382) on the family cache, [ATs & IdTokens will not be returned](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/iambmelt/913/common/src/main/java/com/microsoft/identity/common/internal/cache/MicrosoftFamilyOAuth2TokenCache.java#L144-L154).

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MicrosoftFamilyOAuth2TokenCache.java [hunk 1]
```diff
@@ -91,107 +85,75 @@ public ICacheRecord loadByFamilyId(@Nullable final String clientId,
...
-            final List<RefreshTokenRecord> familyRefreshTokens = new ArrayList<>();
+        // Iterate over those refresh tokens and see if any are in the family...
+        final List<RefreshTokenRecord> familyRefreshTokens = new ArrayList<>();
 
-            for (final RefreshTokenRecord refreshToken : allRefreshTokens) {
-                if (refreshToken.getFamilyId().equals(familyId)) {
-                    familyRefreshTokens.add(refreshToken);
-                }
+        for (final RefreshTokenRecord refreshToken : allRefreshTokens) {
+            if (refreshToken.getFamilyId().equals(familyId)) {
```
Comment (oldalton) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MicrosoftFamilyOAuth2TokenCache.java]: Can we check for familyId and other properties in the for loop above?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MicrosoftFamilyOAuth2TokenCache.java]: I'll collapse these, thanks

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MicrosoftFamilyOAuth2TokenCache.java]: Resolved in:
- 47954879d4df738d726321a9a8110a29669f0954
- abec7f902e22f1e2ad4772c80f4d2d992d1ba2b7

**Comment by iambmelt** [comment]
@shoatman Synced offline. There will only be 1 FRT for the `AccountRecord` so it's always the latest

**Comment by iambmelt** [comment]
Feedback resolved, thanks @oldalton @shoatman

File: common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java [hunk 2]
```diff
@@ -198,7 +198,8 @@ public ICacheRecord save(@NonNull AccountRecord accountRecord,
             // Save to the processUid cache... or create a new one
```
Comment (oldalton) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: I think we shouldn't be even saving access and id token records to foci cache, right? How will id and access tokens be saved to uid based cache if it's a foci token?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java]: Apps will use _either_ the foci cache or the uid cache, but not both.

If an app is not FoCI:
- The `uid`-based cache is used to persist all tokens (+accounts)

If an app is FoCI:
- The `foci` cache is used to store all tokens (+accounts)
- When the presented `uid` matches the `BrokerApplicationMetadata` all tokens (for the requesting account) are returned
- When the presented `uid` _doesn't_ match, only the RT and Account are returned
    * The controller above will then see that no AT was returned and make a `/token` request
    * If the request is successful, the controller will call `save()` and a `BrokerApplicationMetadata` entry will be created
    * If the request is not successful, an the error is wrapped in an Exception and returned to the client

**Pull request merged**